compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs

```diff
@@ -685,6 +672,90 @@ impl<'f> Context<'f> {
         }
     }
 
+    /// Given the arguments of a constrain instruction, multiplying them by the branch's condition
+    /// requires special handling in the case of complex types.
+    fn handle_constrain_arg_side_effects(
+        &mut self,
+        argument: ValueId,
+        condition: ValueId,
+        call_stack: &CallStack,
+    ) -> ValueId {
+        let argument_type = self.inserter.function.dfg.type_of_value(argument);
+
+        match &argument_type {
+            Type::Numeric(_) => {
+                // Condition needs to be cast to argument type in order to multiply them together.
+                let casted_condition = self.insert_instruction(
+                    Instruction::Cast(condition, argument_type),
+                    call_stack.clone(),
+                );
+
+                self.insert_instruction(
+                    Instruction::binary(BinaryOp::Mul, argument, casted_condition),
+                    call_stack.clone(),
+                )
+            }
+            Type::Array(_, _) => {
+                self.handle_array_constrain_arg(argument_type, argument, condition, call_stack)
+            }
+            Type::Slice(_) => {
+                panic!("Cannot use slices directly in a constrain statement")
+            }
+            Type::Reference(_) => {
+                panic!("Cannot use references directly in a constrain statement")
+            }
+            Type::Function => {
+                panic!("Cannot use functions directly in a constrain statement")
+            }
+        }
+    }
+
+    fn handle_array_constrain_arg(
+        &mut self,
+        typ: Type,
+        argument: ValueId,
+        condition: ValueId,
+        call_stack: &CallStack,
+    ) -> ValueId {
+        let mut new_array = im::Vector::new();
+
+        let (element_types, len) = match &typ {
+            Type::Array(elements, len) => (elements, *len),
+            _ => panic!("Expected array type"),
+        };
+
+        for i in 0..len {
+            for (element_index, element_type) in element_types.iter().enumerate() {
+                let index = ((i * element_types.len() + element_index) as u128).into();
+                let index = self.inserter.function.dfg.make_constant(index, Type::field());
+
+                let typevars = Some(vec![element_type.clone()]);
+
+                let mut get_element = |array, typevars| {
+                    let get = Instruction::ArrayGet { array, index };
+                    self.inserter
+                        .function
+                        .dfg
+                        .insert_instruction_and_results(
+                            get,
+                            self.inserter.function.entry_block(),
+                            typevars,
+                            CallStack::new(),
+                        )
+                        .first()
+                };
```

Comment:
We have a `insert_array_get` method which I think could make this cleaner. You may need to switch to the entry block and back again if we're not already on it though.

Reply:
`insert_array_get` is part of the `FunctionBuilder` while we are modifying the DFG directly here. Unless I am missing something?